py/dml/ctree.py

```diff
@@ -4297,6 +4307,191 @@ def read(self):
+    node = templates_subref.templates_ref.node
+    indices = templates_subref.templates_ref.indices
+
+    template_name = templates_subref.template_name
+
+    rank_to_candidate = {}
+
+    # Add any shared method implementation as a possible candidate
+    trait = dml.globals.traits.get(template_name)
+    shared_method = None
+    if trait is not None:
+        for impl_trait in trait.method_impl_traits.get(method_name, []):
+            impl_template = dml.globals.templates[impl_trait.name]
+            shared_method = impl_trait.method_impls[method_name]
+            rank_to_candidate[impl_template.spec.rank] = (
+                impl_template, shared_method)
+
+
+    studied_templates = set()
+    pending_templates = Set((dml.globals.templates[template_name],)
+                            # don't bother going through regular methods
+                            # if if some shared method impl. is
+                            # unoverridable (can only be one if so)
+                            * (shared_method is None
+                               or shared_method.overridable))
+
+    # Find the highest ranking candidate of each distinct template hierarchy
+    # that is actually implemented for the object
+    while pending_templates:
+        tmpl = pending_templates.pop()
+        studied_templates.add(tmpl)
+        (final, implements_method,
+         next_templates) = tmpl.get_potential_method_impl_details(method_name)
+
+        if final:
+            rank_to_candidate.clear()
+            if next_templates:
+                [tmpl] = next_templates
+            elif not implements_method:
+                break
+
+            candidate = node.template_method_impls.get((tmpl, method_name))
+            assert candidate is not None
+            rank_to_candidate[tmpl.spec.rank] = (tmpl, candidate)
+            break
+
+        if implements_method:
+            candidate = node.template_method_impls.get((tmpl, method_name))
+            if candidate is not None:
+                rank_to_candidate[tmpl.spec.rank] = (tmpl, candidate)
+                continue
+
+        pending_templates.update(filter(lambda c: c not in studied_templates,
+                                        next_templates))
+
+
+    # Eliminate any candidates with lower rank of those found
+    # Needed because distinct template hierarchies may still share tails.
+    # E.g. you can have
+    # T -> X (doesn't provide impl. due to #if) -> Z (provides impl.)
+    # T -> Y (provides impl.) -> Z
+    # ... in which case the candidate implementation of Z should be excluded
+    from .traits import calc_minimal_ancestry
+    candidates = [rank_to_candidate[r]
+                  for r in calc_minimal_ancestry(Set(rank_to_candidate))[None]]
+
+    if not candidates:
+        raise EMEMBER(site, templates_subref, method_name)
+
+    if len(candidates) > 1:
+        raise EAMBTQMIC(site, template_name, method_name, candidates)
```

Comment:
Fun fact: I originally had these `raise`s as `report`s, but that caused these errors to be reported multiple times because of `try_codegen_invocation`, and I figured that was not worth it.

Reply:
Indeed, `raise` often gives "remove current object and continue" as fallback, which is good enough for our purposes (to combine many error tests in one file).